lib/JSON/Validator.pm

```diff
@@ -574,69 +573,101 @@ sub _validate {
 
 sub _validate_all_of {
   my ($self, $data, $path, $rules) = @_;
-  my $type = data_type $data, $rules;
-  my (@errors, @expected);
+  my (@errors);
 
   my $i = 0;
   for my $rule (@$rules) {
     next unless my @e = $self->_validate($_[1], $path, $rule);
-    my $schema_type = schema_type $rule;
-    push @expected, $schema_type if $schema_type;
-    push @errors, [$i, @e] if !$schema_type or $schema_type eq $type;
+    push @errors, [$i, @e];
   }
   continue {
     $i++;
   }
 
-  return E $path, [allOf => type => join('/', uniq @expected), $type]
-    if !@errors and @expected;
-  return prefix_errors allOf => @errors if @errors;
-  return;
+  return if not @errors;
+
+  # combine all 'type' errors together
+  return E $path,
+    [
+    allOf => type => join('/', uniq map $_->[1]->details->[0], @errors),
+    $errors[-1][1]->details->[2]
+    ]
+    if @errors > 1 and not grep $_->[1]->details->[1] ne 'type', @errors;
+
+  return prefix_errors allOf => @errors;
+}
+
+sub _validate_any_of_types {
+  my ($self, $data, $path, $rules) = @_;
+  my @errors;
+
+  for my $rule (@$rules) {
+    my @e = $self->_validate($_[1], $path, $rule);
+    return unless @e;
+    push @errors, @e;
+  }
+
+  if (my @e = grep $_->details->[1] ne 'type', @errors) {
+    return @e;
+  }
+
+  # combine all 'type' errors together
+  my $want_types = join('/', uniq map $_->details->[0], @errors);
+  return E $path, [$want_types => 'type', $errors[-1]->details->[2]];
```

Comment:
I would like to use `@details` in the code above:

```
  my @details = grep { $_->[1] eq 'type' } map { $_->details } @errors;
  return @errors unless @details == @errors;

  my $want_types = join '/', map { $_->[0] } @details;
  return E $path, [$want_types => type => $details[-1][2]];
```

Reply:
unfortunately that breaks some tests.  I don't see a good way of shortening the code without changing the logic.